File: Day2.hs
{-# LANGUAGE RecordWildCards #-}
{-# OPTIONS_GHC -Wall #-}

module Day2 where

import Control.Monad.State

data Program =
  Program
    { pOps :: [Int]
    , pIp :: Int
    }
  deriving (Show)

split :: Eq a => a -> [a] -> [[a]]
split s l = go s l [] []
  where
    go :: Eq a => a -> [a] -> [a] -> [[a]] -> [[a]]
    go _ [] buf results = reverse $ buf : results
    go sep (x:xs) buf results
      | sep == x = go sep xs [] (reverse buf : results)
      | otherwise = go sep xs (x : buf) results

parse :: String -> [Int]
parse source = read <$> split ',' source

update :: Int -> a -> [a] -> [a]
update _ _ [] = []
update 0 a (_:xs) = a : xs
update pos a (x:xs) = x : update (pos - 1) a xs

loadProgram :: IO Program
loadProgram =
  Program <$> (update 1 12 <$> update 2 2 <$> parse <$> readFile "Day2.txt") <*>
  return 0

main :: IO ()
main = do
  program <- loadProgram
  print program
  let (a, _) = runState eval program
  print a

eval :: State Program Int
eval = do
  op <- fetchCurrentInstruction
  case op of
    1 -> do
      addIID 1 2 3
      moveIp 4
      eval
    2 -> do
      mulIID 1 2 3
      moveIp 4
      eval
    99 -> head <$> gets pOps
    _ -> error "Unrecognized op code"

moveIp :: Int -> State Program ()
moveIp offset = do
  program <- get
  put program {pIp = pIp program + offset}

addIID :: Int -> Int -> Int -> State Program ()
addIID a b c = do
  x <- readIndirect a
  y <- readIndirect b
  z <- readDirect c
  writeDirect z (x + y)

mulIID :: Int -> Int -> Int -> State Program ()
mulIID a b c = do
  x <- readIndirect a
  y <- readIndirect b
  z <- readDirect c
  writeDirect z (x * y)

fetchCurrentInstruction :: State Program Int
fetchCurrentInstruction = do
  Program {..} <- get
  return $ pOps !! pIp

readDirect :: Int -> State Program Int
readDirect offset = do
  Program {..} <- get
  return $ pOps !! (pIp + offset)

readIndirect :: Int -> State Program Int
readIndirect offset = do
  Program {..} <- get
  return $ pOps !! (pOps !! (pIp + offset))

writeDirect :: Int -> Int -> State Program ()
writeDirect dest value = do
  Program {..} <- get
  modify $ \s -> s {pOps = update dest value pOps}

File: Day2.py
class Machine:

    def __init__(self):
        self._ip = 0
        self._tape = []

    def write_direct(self, addr, value):
        self._tape[addr] = value

    def read_direct_offset(self, offset):
        return self._tape[self._ip + offset]

    def read_indirect_offset(self, offset):
        return self._tape[self._tape[self._ip + offset]]

    def read_direct(self, addr):
        return self._tape[addr]

    def read_indirect(self, addr):
        return self._tape[self._tape[addr]]

    def run(self, program, a, b):

        self._ip = 0

        self._tape = program[:]

        self.write_direct(1, a)
        self.write_direct(2, b)

        while self._ip < len(program):
            op = self.read_direct_offset(0)
            if op == 1:
                x = self.read_indirect_offset(1)
                y = self.read_indirect_offset(2)
                z = self.read_direct_offset(3)
                self.write_direct(z, x + y)
                self._ip += 4
            elif op == 2:
                x = self.read_indirect_offset(1)
                y = self.read_indirect_offset(2)
                z = self.read_direct_offset(3)
                self.write_direct(z, x * y)
                self._ip += 4
            elif op == 99:
                return self.read_direct(0)

def parse(path):
    return [
        int(word)
        for word in open(path, "r").read().split(",")
    ]

def solve(tape, inputs, goal):

    e = Machine()

    for (a, b) in inputs:
        if e.run(tape, a, b) == goal:
            return (a, b, 100*a + b)


if __name__ == "__main__":

    code = parse("Day2.txt")

    machine = Machine()

    print("Part 1:")
    print(machine.run(code, 12, 2))

    inputs = [
        (i, j) for i in range(100) for j in range(100)
    ]

    print("Part 2:")
    print(solve(code, inputs, 19690720))

